Lib/typing.py

```diff
@@ -2050,6 +2083,29 @@ class Film(TypedDict):
     return isinstance(tp, _TypedDictMeta)
 
 
+def assert_never(arg: Never, /) -> Never:
+    """Statically assert that a line of code is unreachable.
+
+    Example::
+
+        def int_or_str(arg: int | str) -> None:
+            match arg:
+                case int():
+                    print("It's an int")
+                case str():
+                    print("It's a str")
+                case _:
+                    assert_never(arg)
+
+    If a type checker finds that a call to assert_never() is
+    reachable, it will emit an error.
+
+    At runtime, this throws an exception when called.
+
+    """
+    raise AssertionError("Expected code to be unreachable")
```

Comment:
Have you given any thought to potentially making the runtime error message user-customisable? You could have something like this:

```python
def assert_never(arg: Never, /, *, msg=None) -> Never:
    if msg is None:
        msg = "Expected code to be unreachable"
    raise AssertionError(msg)
```

Reply:
I don't want to complicate this thing prematurely. I'd rather we land this change so we can move forward with putting it in typing-extensions and type checkers.

Reply:
Sure 👍